Allow to pass keyword arguments to itensor mps estimator for fast computation.

terasakisatoshi opened the pull request:
# Description

This PR allows users to pass keyword arguments to `jl.apply` in [the auxiliary function _estimate](https://github.com/QunaSys/quri-parts/blob/8e370584aeb0d3eb4b0ef3e7049ab9f98e01a48c/packages/itensor/quri_parts/itensor/estimator.py#L44-L62). Note that `jl.apply` Julia function [ITensors.apply](https://itensor.github.io/ITensors.jl/dev/MPSandMPO.html#Gate-evolution). By passing `maxdim`, it enables us to calculate the expectation value at high speed by obtaining the Truncated SVD in [ITensors.factorize](https://github.com/ITensor/ITensors.jl/blob/73bc07eded9bd5f88da1a1562986ac5fb6e42a10/src/tensor_operations/matrix_decomposition.jl#L580-L694) and [svd](https://github.com/ITensor/ITensors.jl/blob/73bc07eded9bd5f88da1a1562986ac5fb6e42a10/NDTensors/src/linearalgebra/linearalgebra.jl#L93C1-L258). These functions are called from the `ITensors.apply`.

- [x] I have added tests that verify the behavior of the changes I made.

terasakisatoshi committed:
pass kwargs to jl.apply to obtain an approximate SVD in ITensor.jl

terasakisatoshi committed:
write test

terasakisatoshi commented:
I have read the CLA Document and I hereby sign the CLA

terasakisatoshi committed:
format

terasakisatoshi committed:
poetry run docformatter -i -r .

toru4838 commented on packages/itensor/quri_parts/itensor/estimator.py:
```diff
@@ -56,17 +58,25 @@ def _estimate(operator: Estimatable, state: ITensorStateT) -> Estimate[complex]:
 
 
-def create_itensor_mps_estimator() -> QuantumEstimator[ITensorStateT]:
+def create_itensor_mps_estimator(**kwargs: Any) -> QuantumEstimator[ITensorStateT]:
```
I think it's better if some commonly used arguments are given explicitly so that IDEs can provide the code suggestions.

```suggestion
def create_itensor_mps_estimator(
    mindim: Optional[int] = None,
    maxdim: Optional[int] = None,
    cutoff: Optional[float] = None,
    **kwargs: Any,
) -> QuantumEstimator[ITensorStateT]:
```

toru4838 commented on packages/itensor/quri_parts/itensor/estimator.py:
```diff
@@ -56,17 +58,25 @@ def _estimate(operator: Estimatable, state: ITensorStateT) -> Estimate[complex]:
+    """
+
+    def estimator(operator: Estimatable, state: ITensorStateT) -> Estimate[complex]:
+        return _estimate(operator, state, **kwargs)
```
There might be a better implementation though
```suggestion
    Args:
        mindim: The minimum number of singular values. The value is passed to `ITensors.apply`.
        maxdim: ...
        cutoff: ...
    Keyword arguments are passed to `ITensors.apply

    <https://itensor.github.io/ITensors.jl/dev/MPSandMPO.html#ITensors.product-Tuple{ITensor,%20ITensors.AbstractMPS}>`_
    """

    def estimator(operator: Estimatable, state: ITensorStateT) -> Estimate[complex]:
        if mindim:
            kwargs["mindim"] = mindim
        if ...
        
        return _estimate(operator, state, **kwargs)
```

terasakisatoshi replied:
Thank you. This is better than mine.

toru4838 reviewed (commented):
Thank you!
Please check my comments. I think the updates here can also be applied to [Sampler](https://github.com/QunaSys/quri-parts/blob/main/packages/itensor/quri_parts/itensor/sampler.py).  Can you do the same for it?

terasakisatoshi committed:
Update packages/itensor/quri_parts/itensor/estimator.py

Co-authored-by: Toru Shibamiya <49753430+toru4838@users.noreply.github.com>

terasakisatoshi commented:
> I think the updates here can also be applied to [Sampler](https://github.com/QunaSys/quri-parts/blob/main/packages/itensor/quri_parts/itensor/sampler.py). Can you do the same for it?

Sure, leave it to me!

terasakisatoshi committed:
Update packages/itensor/quri_parts/itensor/estimator.py

Co-authored-by: Toru Shibamiya <49753430+toru4838@users.noreply.github.com>

terasakisatoshi committed:
Fixed incomplete code in itensor/estimator.py

terasakisatoshi committed:
pass kwargs to jl.apply to obtain an approximate SVD in ITensor.jl

terasakisatoshi committed:
add a comment

terasakisatoshi commented on packages/itensor/quri_parts/itensor/estimator.py:
```diff
@@ -56,17 +58,40 @@ def _estimate(operator: Estimatable, state: ITensorStateT) -> Estimate[complex]:
 
 
-def create_itensor_mps_estimator() -> QuantumEstimator[ITensorStateT]:
+def create_itensor_mps_estimator(
```
It's better to  add `*` so that users can only pass [keyword-only arguments](https://peps.python.org/pep-3102/)?

I mean something like this:

```python
def create_itensor_mps_estimator(
    *,
    mindim: Optional[int] = None,
    maxdim: Optional[int] = None,
    cutoff: Optional[float] = None,
    **kwargs: Any,
) -> QuantumEstimator[ITensorStateT]:
```

terasakisatoshi replied:
fixed in 
30f2833

terasakisatoshi committed:
add some tests to check keyword arguments can be passed to `create_itensor_mps_sampler`

terasakisatoshi commented on packages/itensor/quri_parts/itensor/sampler.py:
```diff
@@ -15,21 +15,47 @@
+    maxdim: Optional[int] = None,
+    cutoff: Optional[float] = None,
+    **kwargs: Any,
+) -> Sampler:
```
Same as https://github.com/QunaSys/quri-parts/pull/203/files/b2ad845f2af51a5ac33c20be28a24e050559c608#r1312476993

terasakisatoshi replied:
30f2833

terasakisatoshi committed:
fix docstring

terasakisatoshi committed:
fix docstring

terasakisatoshi committed:
Enforce that users can only pass keyword-only arguments

terasakisatoshi committed:
remove `mindim` from parameters as it is not often used

Of course we can pass `mindim` as long as ITensors.apply supports it internally.

terasakisatoshi committed:
resolve error regarding flake8

terasakisatoshi committed:
allow to pass keyword arguments e.g., maxdim and cutoff to functions that create a concurrent estimator

terasakisatoshi committed:
allow to pass keywords e.g., maxdim, cutoff to `create_itensor_mps_concurrent_sampler`

terasakisatoshi committed:
add test to check that we can pass keyword arguments

terasakisatoshi committed:
fix docstring

terasakisatoshi committed:
format code in accordance with CONTRIBUTING.md

terasakisatoshi commented:
> Sure, leave it to me!

Done. Ready for review 🙏

toru4838 commented on packages/itensor/quri_parts/itensor/estimator.py:
```diff
@@ -56,34 +58,50 @@ def _estimate(operator: Estimatable, state: ITensorStateT) -> Estimate[complex]:
+    psi = jl.apply(circuit, psi, **kwargs)
     exp: float = jl.expectation(psi, op)
 
     return _Estimate(value=exp, error=0.0)
```
Now `error` should not be fixed to `0.0` since estimate value can have an error if `maxdim` or `cutoff` is set.  It would be nice if you can calculate theoretical value for it (but seems complicated… it may depends on `n_qubits`,  `state`, and so on).  
You can set it's value to`error=np.nan`

terasakisatoshi replied:
> You can set it's value to `error=np.nan`

I've selected the option.

toru4838 commented on packages/itensor/quri_parts/itensor/estimator.py:
```diff
@@ -56,34 +58,50 @@ def _estimate(operator: Estimatable, state: ITensorStateT) -> Estimate[complex]:
+    arguments `**kwargs` are passed to `ITensors.apply.
 
-    return _estimate
+    <https://itensor.github.io/ITensors.jl/dev/MPSandMPO.html#ITensors.product-Tuple{ITensor,%20ITensors.AbstractMPS}>`_
```
This part is not rendered correctly.

toru4838 commented on packages/itensor/quri_parts/itensor/estimator.py:
```diff
@@ -148,15 +171,40 @@ def create_itensor_mps_concurrent_estimator(
+
+    The following parameters including
+    keyword arguments `**kwargs` are passed to `ITensors.apply
+    <https://itensor.github.io/ITensors.jl/dev/MPSandMPO.html#ITensors.product-Tuple{ITensor,%20ITensors.AbstractMPS}>`_
```
```suggestion
    keyword arguments `**kwargs` are passed to `ITensors.apply
    <https://itensor.github.io/ITensors.jl/dev/MPSandMPO.html#ITensors.product-Tuple{ITensor,%20ITensors.AbstractMPS}>`_.
```

toru4838 commented on packages/itensor/quri_parts/itensor/estimator.py:
```diff
@@ -169,31 +217,82 @@ def estimator(
+
+    The following parameters including
+    keyword arguments `**kwargs` are passed to `ITensors.apply
+    <https://itensor.github.io/ITensors.jl/dev/MPSandMPO.html#ITensors.product-Tuple{ITensor,%20ITensors.AbstractMPS}>`_
```
```suggestion
    <https://itensor.github.io/ITensors.jl/dev/MPSandMPO.html#ITensors.product-Tuple{ITensor,%20ITensors.AbstractMPS}>`_.
```

toru4838 commented on packages/itensor/quri_parts/itensor/estimator.py:
```diff
@@ -169,31 +217,82 @@ def estimator(
+
+    The following parameters including
+    keyword arguments `**kwargs` are passed to `ITensors.apply
+    <https://itensor.github.io/ITensors.jl/dev/MPSandMPO.html#ITensors.product-Tuple{ITensor,%20ITensors.AbstractMPS}>`_
```
```suggestion
    <https://itensor.github.io/ITensors.jl/dev/MPSandMPO.html#ITensors.product-Tuple{ITensor,%20ITensors.AbstractMPS}>`_.
```

toru4838 reviewed (commented):
Thank you!
I have made a few minor comments. Please take care of it 🙏

terasakisatoshi committed:
Update packages/itensor/quri_parts/itensor/estimator.py

add `.`

Co-authored-by: Toru Shibamiya <49753430+toru4838@users.noreply.github.com>

terasakisatoshi committed:
Update packages/itensor/quri_parts/itensor/estimator.py

Co-authored-by: Toru Shibamiya <49753430+toru4838@users.noreply.github.com>

terasakisatoshi committed:
Update packages/itensor/quri_parts/itensor/estimator.py

Co-authored-by: Toru Shibamiya <49753430+toru4838@users.noreply.github.com>

terasakisatoshi committed:
set `error = np.nan` if "maxdim" in kwargs or "cutoff" in kwargs

terasakisatoshi committed:
add tests

toru4838 commented on packages/itensor/quri_parts/itensor/estimator.py:
```diff
@@ -56,41 +58,68 @@ def _estimate(operator: Estimatable, state: ITensorStateT) -> Estimate[complex]:
+    calculate expectation values. The following parameters including keyword
+    arguments `**kwargs` are passed to `ITensors.apply.
+
+    <https://itensor.github.io/ITensors.jl/dev/MPSandMPO.html#ITensors.product-Tuple{ITensor,%20ITensors.AbstractMPS}>`_
```
```suggestion
    calculate expectation values. The following parameters including keyword
    arguments `**kwargs` are passed to `ITensors.apply
    <https://itensor.github.io/ITensors.jl/dev/MPSandMPO.html#ITensors.product-Tuple{ITensor,%20ITensors.AbstractMPS}>`_.
```

toru4838 commented on packages/itensor/quri_parts/itensor/estimator.py:
```diff
@@ -148,15 +182,40 @@ def create_itensor_mps_concurrent_estimator(
     >>> with ProcessPoolExecutor(
     ...     max_workers=2, mp_context=get_context("spawn")
     ... ) as executor:
```
This is not your work but now it is the right time to fix 👍
```suggestion
    Examples:
        >>> with ProcessPoolExecutor(
                max_workers=2, mp_context=get_context("spawn")
            ) as executor:
```

toru4838 reviewed (changes requested):
Almost!

terasakisatoshi committed:
Update packages/itensor/quri_parts/itensor/estimator.py

Co-authored-by: Toru Shibamiya <49753430+toru4838@users.noreply.github.com>

terasakisatoshi committed:
Update packages/itensor/quri_parts/itensor/estimator.py

Co-authored-by: Toru Shibamiya <49753430+toru4838@users.noreply.github.com>

toru4838 commented on packages/itensor/quri_parts/itensor/estimator.py:
```diff
@@ -56,41 +58,67 @@ def _estimate(operator: Estimatable, state: ITensorStateT) -> Estimate[complex]:
-    calculate expectation values."""
+    calculate expectation values. The following parameters including keyword
+    arguments `**kwargs` are passed to `ITensors.apply
+    <https://itensor.github.io/ITensors.jl/dev/MPSandMPO.html#ITensors.product-Tuple{ITensor,%20ITensors.AbstractMPS}>`_.
```
```suggestion
    calculate expectation values.
    
    The following parameters including keyword
    arguments `**kwargs` are passed to `ITensors.apply
    <https://itensor.github.io/ITensors.jl/dev/MPSandMPO.html#ITensors.product-Tuple{ITensor,%20ITensors.AbstractMPS}>`_.
```

toru4838 commented on packages/itensor/quri_parts/itensor/estimator.py:
```diff
@@ -138,25 +166,56 @@ def _concurrent_estimate(
     """Returns a :class:`~ConcurrentQuantumEstimator` that uses ITensor MPS
     simulator to calculate expectation values.
 
     For now, this function works when the executor is defined like below::
```
```suggestion
    For now, this function works when the executor is defined like below
```

toru4838 reviewed (changes requested):
Sorry the documentation is still broken.

terasakisatoshi committed:
Update packages/itensor/quri_parts/itensor/estimator.py

Co-authored-by: Toru Shibamiya <49753430+toru4838@users.noreply.github.com>

terasakisatoshi committed:
Update packages/itensor/quri_parts/itensor/estimator.py

Co-authored-by: Toru Shibamiya <49753430+toru4838@users.noreply.github.com>

toru4838 commented on packages/itensor/quri_parts/itensor/estimator.py:
```diff
@@ -56,41 +58,69 @@ def _estimate(operator: Estimatable, state: ITensorStateT) -> Estimate[complex]:
+) -> QuantumEstimator[ITensorStateT]:
+    """Returns a :class:`~QuantumEstimator` that uses ITensor MPS simulator to
+    calculate expectation values.
+    
```
I hope this is the final modification 🙏
```suggestion

```

terasakisatoshi replied:
Thank you!

terasakisatoshi committed:
Update packages/itensor/quri_parts/itensor/estimator.py

Co-authored-by: Toru Shibamiya <49753430+toru4838@users.noreply.github.com>

toru4838 commented on packages/itensor/quri_parts/itensor/estimator.py:
```diff
@@ -56,41 +58,68 @@ def _estimate(operator: Estimatable, state: ITensorStateT) -> Estimate[complex]:
+    calculate expectation values.
+    The following parameters including keyword
+    arguments `**kwargs` are passed to `ITensors.apply
+    <https://itensor.github.io/ITensors.jl/dev/MPSandMPO.html#ITensors.product-Tuple{ITensor,%20ITensors.AbstractMPS}>`_.
```
My last suggestion was not correct…
What I wanted is to insert blank line between first and second sentence, with no whitespaces.
```suggestion
    """Returns a :class:`~QuantumEstimator` that uses ITensor MPS simulator to
    calculate expectation values.

    The following parameters including keyword
    arguments `**kwargs` are passed to `ITensors.apply
    <https://itensor.github.io/ITensors.jl/dev/MPSandMPO.html#ITensors.product-Tuple{ITensor,%20ITensors.AbstractMPS}>`_.
```

terasakisatoshi replied:
👌

terasakisatoshi committed:
Update packages/itensor/quri_parts/itensor/estimator.py

Co-authored-by: Toru Shibamiya <49753430+toru4838@users.noreply.github.com>

toru4838 reviewed (approved):
Great work!
